### PR TITLE
[IOTDB-5557] [RatisConsensus] Support linearizable read during recovery

### PR DESCRIPTION
**Pull request by SzyWilliam**
The default level is set to "query statemachine directly”, which, while latency-friendly, has led to user-reported bugs. Specifically, these bugs relate to the production of inconsistent results in subsequent SQL queries during a restart, creating a phantom read problem that may be confusing for our users.

To address this issue, I propose that we temporarily increase the read consistency level to linearizable read during restarts. This will ensure that we maintain data consistency during the critical recovery period. Once the cluster has successfully finished recovering from previous logs, we can then revert to the default consistency level.

### CODE REVIEW

**Comment by lancelly** [comment]
> The default level is set to "query statemachine directly”, which, while latency-friendly, has led to user-reported bugs. Specifically, these bugs relate to the production of inconsistent results in subsequent SQL queries during a restart, creating a phantom read problem that may be confusing for our users.
> 
> To address this issue, I propose that we temporarily increase the read consistency level to linearizable read during restarts. This will ensure that we maintain data consistency during the critical recovery period. Once the cluster has successfully finished recovering from previous logs, we can then revert to the default consistency level.

LGTM~ Song is true master of Ratis indeed🤩

**Review by CRZbulabula — APPROVED** [review]
LGTM! Let me introduce William Song, a shining star in both Apache IoTDB and Apache Ratis community.

**Comment by Cpaulyz** [comment]
Excellent work! Thank you for working day and night! This is a great contribution for IoTDB. 🤟💘

**Comment by BUAAserein** [comment]
Awesome work! Thanks William Song @SzyWilliam very much for working on this! 🤟💘

File: iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/ratis/RatisConsensus.java [hunk 1]
```diff
@@ -803,6 +838,11 @@ public RaftServer getServer() {
     return server;
   }
 
+  @TestOnly
+  public void allowStaleRead(ConsensusGroupId consensusGroupId) {
+    canServeStaleRead.computeIfAbsent(consensusGroupId, id -> new AtomicBoolean(true));
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/ratis/RatisConsensus.java:843]: canServeStaleRead.computeIfAbsent(consensusGroupId, id -> new AtomicBoolean(true)).set(true) ?

File: iotdb-core/consensus/src/test/java/org/apache/iotdb/consensus/ratis/RecoverReadTest.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+    miniCluster.getStateMachines().stream()
+        .map(SlowRecoverStateMachine.class::cast)
+        .forEach(SlowRecoverStateMachine::setStall);
+
+    // shutdown and restart the cluster
+    miniCluster.restart();
+
+    // query during redo: get exception that ratis is under recovery
+    final ConsensusReadResponse readResponse = TestUtils.doRead(miniCluster.getServer(0), gid);
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/consensus/src/test/java/org/apache/iotdb/consensus/ratis/RecoverReadTest.java:173]: The execution time for 10 logs is 5s, and the query timeout is 30s. Why is there an exception?

**Review by OneSizeFitsQuorum — COMMENTED** [review]
Awesome work！
Although the Ratis community does not currently support lease read, this work has allowed us to avoid a lot of unwanted anomalies.
I was wondering if we could avoid this kind of anomalies when the new leader takes over, so that some of the upper confingode modules don't see the old state. I think RatisConsensus and ApplicationStateMachineProxy of Boolean variables can be shared to meet this requirement

File: iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/ratis/RatisConsensus.java (hunk 1 above)
Comment (SzyWilliam) [reply iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/ratis/RatisConsensus.java]: Good catch!

File: iotdb-core/consensus/src/test/java/org/apache/iotdb/consensus/ratis/RecoverReadTest.java (hunk 2 above)
Comment (SzyWilliam) [reply iotdb-core/consensus/src/test/java/org/apache/iotdb/consensus/ratis/RecoverReadTest.java:178]: It is caused by no-leader. No active leader will result in immediate return of failure.

**Comment by SzyWilliam** [comment]
@lancelly @Cpaulyz @CRZbulabula @BUAAserein Thanks a lot for reviewing at this PR! You are the true heroes of IoTDB Community!

**Comment by SzyWilliam** [comment]
@OneSizeFitsQuorum Thanks for the careful review! Made corresponding changes on code. PTAL.
How about keep this PR to 'restart' and propose 'leader-switch' related changes in a new PR?

**Comment by OneSizeFitsQuorum** [comment]
> @OneSizeFitsQuorum Thanks for the careful review! Made corresponding changes on code. PTAL. How about keep this PR to 'restart' and propose 'leader-switch' related changes in a new PR?

Good Idea！After fixing suspected errors in CI, I will merge this PR. Looking forward to your next PR!

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

**Pull request merged**